File: tests/oracle/funcproto_lock_1204.cc
#include <iostream>

#include "fspec.hh"

using namespace ghidra;

static void dump(const char *label, FuncProto &proto) {
  std::cout << label
            << ":input=" << proto.isInputLocked()
            << ",output=" << proto.isOutputLocked()
            << ",model=" << proto.isModelLocked()
            << ",params=" << proto.numParams() << '\n';
}

int main() {
  TypeVoid void_type;
  FuncProto proto;
  proto.setInternal(nullptr, &void_type);
  dump("fresh", proto);

  proto.setInputLock(true);
  dump("input_locked", proto);
  proto.clearUnlockedInput();
  dump("clear_unlocked", proto);

  FuncProto copied;
  copied.copy(proto);
  dump("copied", copied);
  copied.clearInput();
  dump("clear_input", copied);

  FuncProto output;
  output.setInternal(nullptr, &void_type);
  output.setOutputLock(true);
  dump("output_locked", output);
  output.setOutputLock(false);
  dump("output_unlocked", output);
  return 0;
}

File: tests/oracle/funcproto_lock_1204.rs
use std::sync::Arc;

use rugra::fspec::FuncProto;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeMetatype};

fn void_type() -> Arc<Datatype> {
    Arc::new(Datatype::Void(TypeBase::new(
        "void".to_string(),
        0,
        TypeMetatype::Void,
    )))
}

fn dump(label: &str, proto: &FuncProto) {
    println!(
        "{label}:input={},output={},model={},params={}",
        u8::from(proto.is_input_locked()),
        u8::from(proto.is_output_locked()),
        u8::from(proto.is_model_locked()),
        proto.num_params(),
    );
}

fn main() {
    let void_type = void_type();
    let mut proto = FuncProto::new(String::new(), void_type.clone());
    dump("fresh", &proto);

    proto.set_input_lock(true);
    dump("input_locked", &proto);
    proto.clear_unlocked_input();
    dump("clear_unlocked", &proto);

    let mut copied = FuncProto::new(String::new(), void_type.clone());
    copied.copy_from(&proto);
    dump("copied", &copied);
    copied.clear_input();
    dump("clear_input", &copied);

    let mut output = FuncProto::new(String::new(), void_type);
    output.set_output_lock(true);
    dump("output_locked", &output);
    output.set_output_lock(false);
    dump("output_unlocked", &output);
}
